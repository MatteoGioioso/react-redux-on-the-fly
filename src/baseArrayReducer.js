/**
 * BaseArrayReducerMethods
 */
export class BaseArrayReducerMethods {
  _update(index, value, array) {
    return [...array.slice(0, index), value, ...array.slice(index + 1)];
  }

  receiveMany(state, payload) {
    return [...state, ...payload];
  }

  receiveOne(state, payload) {
    return [...state, payload];
  }

  substituteOne(state, payload) {
    const indexOfPayload = state.findIndex(item => item.id === payload.id);
    return this.update(indexOfPayload, payload, state);
  }

  destroyOne(state, payload) {
    return state.filter(item => item.id !== payload);
  }

  resetAll(state, payload) {
    return [];
  }
}

export const baseArrayReducer = (
  section,
  initialState = [],
  Overrides = BaseArrayReducerMethods,
  reducer
) => (state = initialState, action) => {
  const methods = new Overrides();

  switch (action.type) {
    case `RECEIVE_${section}S`:
      return methods.receiveMany(state, action.payload);
    case `RECEIVE_${section}`:
      return methods.receiveOne(state, action.payload);
    case `SUBSTITUTE_${section}`:
      return methods.substituteOne(state, action.payload);
    case `DESTROY_${section}`:
      return methods.destroyOne(state, action.payload);
    case `RESET_${section}S`:
      return methods.resetAll(state, action.payload);
    default:
      return reducer ? reducer(state, action) : state;
  }
};
