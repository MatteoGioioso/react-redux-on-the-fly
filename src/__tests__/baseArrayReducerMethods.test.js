import { BaseArrayReducerMethods } from "../baseArrayReducer";

describe("BaseArrayReducerMethods class", function() {
  let state;

  beforeEach(function() {
    state = [
      { id: "1", content: "stuff_1" },
      { id: "2", content: "stuff_2" },
      { id: "3", content: "stuff_3" },
      { id: "4", content: "stuff_4" },
      { id: "5", content: "stuff_5" },
      { id: "6", content: "stuff_6" },
      { id: "7", content: "stuff_7" },
      { id: "8", content: "stuff_8" },
      { id: "9", content: "stuff_9" },
      { id: "10", content: "stuff_10" }
    ];
  });

  it("should destroy multiple elements in array", function() {
    const payload = ["3", "4", "6"];
    const result = new BaseArrayReducerMethods().destroyMany(state, payload);

    expect(result).toHaveLength(7);
    expect(result).toMatchSnapshot();
  });

  it("should destroy multiple elements in array if one id does not math", function() {
    const payload = ["3", "4", "11"];
    const result = new BaseArrayReducerMethods().destroyMany(state, payload);

    expect(result).toHaveLength(8);
    expect(result).toMatchSnapshot();
  });

  it("should substitute one item in array", function() {
    const payload = { id: "5", content: "stuff_5_updated" };
    const result = new BaseArrayReducerMethods().substituteOne(state, payload);

    expect(result[4]).toEqual(payload);
    expect(result).toMatchSnapshot();
  });

  it("should not substitute a item in array, id does not match", function() {
    const payload = { id: "11", content: "stuff_11_updated" };
    const result = new BaseArrayReducerMethods().substituteOne(state, payload);

    expect(result).toMatchSnapshot();
  });
});
