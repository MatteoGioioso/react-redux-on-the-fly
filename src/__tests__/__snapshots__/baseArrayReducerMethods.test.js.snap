// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseArrayReducerMethods class should destroy multiple elements in array 1`] = `
Array [
  Object {
    "content": "stuff_1",
    "id": "1",
  },
  Object {
    "content": "stuff_2",
    "id": "2",
  },
  Object {
    "content": "stuff_5",
    "id": "5",
  },
  Object {
    "content": "stuff_7",
    "id": "7",
  },
  Object {
    "content": "stuff_8",
    "id": "8",
  },
  Object {
    "content": "stuff_9",
    "id": "9",
  },
  Object {
    "content": "stuff_10",
    "id": "10",
  },
]
`;

exports[`BaseArrayReducerMethods class should destroy multiple elements in array if one id does not math 1`] = `
Array [
  Object {
    "content": "stuff_1",
    "id": "1",
  },
  Object {
    "content": "stuff_2",
    "id": "2",
  },
  Object {
    "content": "stuff_5",
    "id": "5",
  },
  Object {
    "content": "stuff_6",
    "id": "6",
  },
  Object {
    "content": "stuff_7",
    "id": "7",
  },
  Object {
    "content": "stuff_8",
    "id": "8",
  },
  Object {
    "content": "stuff_9",
    "id": "9",
  },
  Object {
    "content": "stuff_10",
    "id": "10",
  },
]
`;

exports[`BaseArrayReducerMethods class should not substitute a item in array, id does not match 1`] = `
Array [
  Object {
    "content": "stuff_1",
    "id": "1",
  },
  Object {
    "content": "stuff_2",
    "id": "2",
  },
  Object {
    "content": "stuff_3",
    "id": "3",
  },
  Object {
    "content": "stuff_4",
    "id": "4",
  },
  Object {
    "content": "stuff_5",
    "id": "5",
  },
  Object {
    "content": "stuff_6",
    "id": "6",
  },
  Object {
    "content": "stuff_7",
    "id": "7",
  },
  Object {
    "content": "stuff_8",
    "id": "8",
  },
  Object {
    "content": "stuff_9",
    "id": "9",
  },
  Object {
    "content": "stuff_10",
    "id": "10",
  },
]
`;

exports[`BaseArrayReducerMethods class should substitute one item in array 1`] = `
Array [
  Object {
    "content": "stuff_1",
    "id": "1",
  },
  Object {
    "content": "stuff_2",
    "id": "2",
  },
  Object {
    "content": "stuff_3",
    "id": "3",
  },
  Object {
    "content": "stuff_4",
    "id": "4",
  },
  Object {
    "content": "stuff_5_updated",
    "id": "5",
  },
  Object {
    "content": "stuff_6",
    "id": "6",
  },
  Object {
    "content": "stuff_7",
    "id": "7",
  },
  Object {
    "content": "stuff_8",
    "id": "8",
  },
  Object {
    "content": "stuff_9",
    "id": "9",
  },
  Object {
    "content": "stuff_10",
    "id": "10",
  },
]
`;
